File: src/Response.py
"""Module containing a class used to represent requests forwarded by a node
that need to be responded to
"""


class Response:
    """Class storing essential information about a frame that has been forwarded
    to correctly interpret responses and send them to requesters as necessary.

    Attributes:
        remaining_responses: the number of nodes responses are expected from
        src: a list of nodes the frame has passed through in chronological order
        origin: the node the frame originated with
        seqno: the sequence number applied to the frame at its origin
        time: the best time received in response to the request yet received
        stop: the node from which the fastest response originated 
    """

    remaining_responses: int
    src: list
    origin: str
    seqno: int
    time: int
    stop: str

    def __init__(
        self, remaining_responses: int, src: list, origin: str, seqno: int, time: int, stop: str
    ):
        """Initialises instance
        """
        self.remaining_responses = remaining_responses
        self.src = src
        self.origin = origin
        self.seqno = seqno
        self.time = time
        self.stop = stop

    def __str__(self):
        """Returns a string representation for easy printing
        """
        return (
            f"Remaining Responses: {self.remaining_responses}\n"
            f"SRC: {self.src}\n"
            f"Origin: {self.origin}\n"
            f"Seqno: {self.seqno}\n"
            f"Best time: {self.time}\n"
            f"Best stop: {self.stop}\n"
        )

File: src/response.cpp
/* 
Implementation of Response class and its methods
*/
#include "response.h"
/*
Constructor for Response objects.

Sets the best time to -1 to ensure any valid time will be recorded. 
*/
Response::Response(int remaining_responses, list<string>& src, string &origin,
                   int seqno)
    : remaining_responses(remaining_responses), src(src), origin(origin),
      seqno(seqno) {
  time = -1;
}
/*
Method to return a string representation of a Response object for easy printing

Arguments:
  void
Returns:
  string representation of Response object
*/
string Response::to_string() {
  ostringstream ss;
  ostringstream src_ss;
  for(auto node: src) {
    src_ss << node << ".";
  }
  ss << "Remaining responses: " << remaining_responses << endl
     << "Frame origin: " << origin << endl
     << "SRC " << src_ss.str() << endl
     << "Seqno: " << seqno << endl
     << "Best time: " << time << endl
     << "Best stop: " << stop << endl;
  return ss.str();
}